### ISSUE TITLE
messaging + react-native-push-notification integration

### ISSUE THREAD

**Issue by alex-mironov** [invertase/react-native-firebase#2997]
After migrating to RN `v.0.61.4` + `react-native-firebase v.6.2.0` I'm trying to make notifications working on Android. Despite the known issues with `getToken` I've managed to make them work on iOS.

On Android, though notifications are not appearing, I assume the issue is somewhere in `Android.xml`. As documentation of **react-native-firebase 6** does not provide specific details for Android and **RN 0.60+** I rely on autolinking with minimum configuration.

Looking into RNFB [documentation](https://github.com/invertase/react-native-firebase-docs/blob/master/docs/messaging/android.md), AndroidManifest.xml should contain

```
  <service android:name="io.invertase.firebase.messaging.RNFirebaseMessagingService">
    <intent-filter>
      <action android:name="com.google.firebase.MESSAGING_EVENT" />
    </intent-filter>
  </service>
```
But as per [react-native-push-notification](https://github.com/zo0r/react-native-push-notification) there should be

```
        <service
            android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationListenerService"
            android:exported="false" >
            <intent-filter>
                <action android:name="com.google.firebase.MESSAGING_EVENT" />
            </intent-filter>
        </service>
```
I'm not really sure which of those versions to use. Would be grateful for any hint.

`package.json`
```
    "@react-native-firebase/app": "^6.2.0",
    "@react-native-firebase/messaging": "^6.2.0",
    "react-native-push-notification": "^3.1.9",
    "react-native": "0.61.4",
```

`android/app/build.gradle`
```
dependencies {
    implementation fileTree(dir: "libs", include: ["*.jar"])
    implementation "com.facebook.react:react-native:+"  // From node_modules
    implementation 'com.google.firebase:firebase-analytics:17.2.0'
    implementation 'com.android.support:multidex:1.0.3'
...
```

[AndroidManifest.txt](https://github.com/invertase/react-native-firebase/files/3955394/AndroidManifest.txt)
[build.gradle.txt](https://github.com/invertase/react-native-firebase/files/3955403/build.gradle.txt)

**Comment by alex-mironov** [comment]
[Documentation](https://github.com/invertase/react-native-firebase-docs/blob/master/docs/messaging/android.md) also suggests updating `android/app/build.gradle` with
```
dependencies {
  // ...
  implementation "com.google.firebase:firebase-messaging:{{ android.firebase.messaging }}"
  implementation 'me.leolin:ShortcutBadger:1.1.21@aar' // <-- Add this line if you wish to use badge on Android
}
```
And registering the package in `MainApplication.java` . I skipped those steps, I assume they're not necessary if I rely on autolinking?

**Comment by mikehardy** [comment]
I think more important than the documentation might be looking at the code - usually each package here is relatively small so you can follow the bootstrap in v6 to see what exactly is auto-linked in.

For example you can see here messaging is already in there: https://github.com/invertase/react-native-firebase/blob/master/packages/messaging/android/build.gradle#L73

Already in the manifest too https://github.com/invertase/react-native-firebase/blob/master/packages/messaging/android/src/main/AndroidManifest.xml#L14

ShortcutBadger is not - partially because it's not a good idea (shortcut badging is not a platform feature, it's a vendor-specific launcher feature that not all launchers implement, which means it's a hack on android and not reliable) - and partially because it's more for local notifications

It's also important to be clear on "messaging" (the push from cloud to device) and "notifications". RNFBv5 handled both. react-native-push-notification can handle both. But if you are using RNFB you want it to handle messaging, and you want react-native-push-notification to handle notifications. So you would register RNFB as the listener for that intent, and just use the other to pop local notifications. That tells you what to register in the manifest

I would carefully examine unfiltered logcat while sending a message. Check https://dontkillmyapp.com for your phone. And just console.log handled cloud messages to make sure you receive data consistently using RNFB alone until that was settled. Then try for react-native-push-notifications

**Comment by mikehardy** [comment]
possibly worth mentioning: if you have anything to share for the people suffering on iOS receiving messages, I'm sure they'd be grateful!

**Comment by alex-mironov** [comment]
@mikehardy thanks for your support, I'm checking it out. 

Talking about iOS, I just delayed the token request as suggested by @Salakar. Here is the [original answer](https://github.com/invertase/react-native-firebase/issues/2657#issuecomment-562174052).

`    const hasPermissions = await messaging().hasPermission()
    if (hasPermissions) {
      await messaging().registerForRemoteNotifications()
      await new Promise((resolve, reject) => setTimeout(() => resolve(), 1000))

      token = await messaging().getToken()
      console.log('FCM token', token)
    } else {
      const { status } = await requestNotifications(['alert', 'sound'])

      if (status === 'granted') {
        await messaging().registerForRemoteNotifications()
        await new Promise((resolve, reject) => setTimeout(() => resolve(), 1000))

        token = await messaging().getToken()
        console.log('FCM token has been received', token)
      }
    }`
It's kinda messy but works for now.

**Comment by alex-mironov** [comment]
@mikehardy can you please clarify this part?

>  So you would register RNFB as the listener for that intent, and just use the other to pop local notifications.

With the workflow you've described, do you have any idea how to combine these 2 configs which use the same intent?

`react-native-push-notification`
```
      <service
        android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationListenerService"
        android:exported="false" >
        <intent-filter>
            <action android:name="com.google.firebase.MESSAGING_EVENT" />
        </intent-filter>
      </service>
```

`react-native-firebase`
```
    <service android:name="io.invertase.firebase.messaging.ReactNativeFirebaseMessagingService"
             android:exported="false">
      <intent-filter>
        <action android:name="com.google.firebase.MESSAGING_EVENT"/>
      </intent-filter>
    </service>
```

**Comment by alex-mironov** [comment]
I end up using 
```
      <receiver android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationPublisher" />
      <receiver android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationBootEventReceiver">
          <intent-filter>
              <action android:name="android.intent.action.BOOT_COMPLETED" />
          </intent-filter>
      </receiver>
      <service android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationRegistrationService"/>

      <service
        android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationListenerService"
        android:exported="false" >
        <intent-filter>
            <action android:name="com.google.firebase.MESSAGING_EVENT" />
        </intent-filter>
      </service>
```
but that doesn't seem to make any difference.

**Comment by nkashyap** [comment]
Hello guys,

I'm also trying to solve a similar issue. for me, Android is working fine but not on ios 13.3.

I have to add the following code in order to request for user permissions.  

```
  if ([UNUserNotificationCenter class] != nil) {
    // iOS 10 or later
    // For iOS 10 display notification (sent via APNS)
    [UNUserNotificationCenter currentNotificationCenter].delegate = self;
    UNAuthorizationOptions authOptions =  UNAuthorizationOptionAlert |
                                          UNAuthorizationOptionSound |
                                          UNAuthorizationOptionBadge;
    [[UNUserNotificationCenter currentNotificationCenter]
      requestAuthorizationWithOptions:authOptions
      completionHandler:^(BOOL granted, NSError * _Nullable error) {
        // ...
      }];
  } else {
    // iOS 10 notifications aren't available; fall back to iOS 8-9 notifications.
    UIUserNotificationType allNotificationTypes =
    (UIUserNotificationTypeSound | UIUserNotificationTypeAlert | UIUserNotificationTypeBadge);
    UIUserNotificationSettings *settings =
    [UIUserNotificationSettings settingsForTypes:allNotificationTypes categories:nil];
    [application registerUserNotificationSettings:settings];
  }

  [application registerForRemoteNotifications];
```

**Comment by mikehardy** [comment]
@nkashyap please don't hijack the thread, this is focused on android. @alex-mironov has ios working. So y ou have a different problem, can you open a new issue?

**Comment by mikehardy** [comment]
You don't want RNPN (xxxpush-notifications - too much to type over and over) to get the firebase cloud messages. RNFB handles that, or should. 

- Send silent-push request to firebase messaging servers somehow
- they blast out an FCM (firebase cloud message) to your devices
- devices have RNFB listening (via auto-link etc - I showed code references above) and they get the silent push
- your code has onMessage called
- in onMessage you just use the local notification APIs of RNPN

In other words, I think you should stop trying to get RNPN to catch cloud messages. You just want to use it for it's local device notification API coverage. That's what RNFBv6 dropped - it doesn't handle local device notifications. But it has cloud messaging covered

**Comment by alex-mironov** [comment]
@mikehardy thanks for the extensive response. 

I get the idea, there are just a couple of issues:
1. I can't get `onMessage` fired. `const fcmToken = await firebase.messaging().getToken()` works and eventually gives me FCM token. But when I try to send a message through the firebase console it goes nowhere. I'm unable to get any log in the simulator, `onMessage` is never hit.
2. From what I understand this wouldn't work when the app is closed, is it correct?

**Comment by mikehardy** [comment]
Not entirely sure. Messaging is hard. Have you looked at the raw logcat of the device as you try it? Have you checked https://dontkillmyapp.com ? Is it a real device or emulator? There are videos and articles on firebase official blog that step through troubleshooting of cloud message delivery. They're gold

**Comment by desis123** [comment]
I have not tried it yet, Tonight I would give it a try. Meanwhile you  can try it   https://github.com/JscramblerBlog/RNnotifications-demo .   Nothing is documented, It the author said it's just working fine.

**Comment by alex-mironov** [comment]
I'm actually using an emulator and real device. There is no luck on either. One thing I've noticed is `Instance ID SDK is deprecated, com.myApp should update to use Firebase Instance ID`

```
2019-12-12 22:55:32.422 11062-11062/com.myApp I/FirebaseInitProvider: FirebaseApp initialization successful
2019-12-12 22:55:32.424 11062-11062/com.myApp D/ReactNativeFirebaseApp: received application context.
2019-12-12 22:55:32.505 11062-11110/com.myApp I/FA: To enable faster debug mode event logging run:
      adb shell setprop debug.firebase.analytics.app com.myApp
2019-12-12 22:55:37.001 11062-11176/com.myApp W/InstanceID: Instance ID SDK is deprecated, com.myApp should update to use Firebase Instance ID
```

Also pretty weird thing, I've downgraded to
```
"@react-native-firebase/messaging": "^6.1.0",
```
and for some time I have been seeing the notifications (not data messages) on emulator after sending them through the firebase console. I know it sounds ridiculous, but then it just stopped working.

**Comment by mikehardy** [comment]
Basically nothing surprises me with regard to failure modes on cloud messaging at this point. So nothing sounds ridiculous either. And there is probably a good reason, which I would add as reason number, maybe 4,652 why messaging might not work. It's maddening. Sorry I don't have better ideas at the moment. If you haven't examined the firebase official blog "cloud messaging troubleshooting" video you might give it a try - they're content is pretty good.

**Comment by alex-mironov** [comment]
Got it, thank you a lot @mikehardy.
Will try to use AWS amplify for push notifications. If it wouldn't work will come back to RNFB...

**Comment by alex-mironov** [comment]
It's interesting, I'm getting different tokens when using `messaging().getToken()` and
```
PushNotification.configure({
      // (optional) Called when Token is generated (iOS and Android)
      onRegister: (token) => {
        console.log('token', token.token)
        AsyncStorage.setItem('token', token.token)
      },
     ....
```
So far I've been using the one received from `messaging().getToken()`. I attempted to use the one received in `onRegister` and apparently notifications are coming, even in Android emulator. 
@mikehardy do you know anything about the difference of those tokens? Probably some possible issues?

**Comment by mikehardy** [comment]
You're much deeper into messaging than I am, we do global pushes in my work project so I haven't messed with tokens much. Always possible documentation needs an update based on current findings

**Comment by GioLogist** [comment]
> It's interesting, I'm getting different tokens when using messaging().getToken() and

`react-native-push-notification` gives you the device token, whereas `messaging().getToken()` gives you an [fcm](https://stackoverflow.com/questions/37671380/what-is-fcm-token-in-firebase) token, specific to Google.

> . I'm unable to get any log in the simulator, onMessage is never hit.

That's odd. I'm having different results (v6.2.0 as well)

On my end, using `react-native-push-notification`: 
- iOS is receiving events from `onNotification` (react-native-push-notification) but not from `onMessage` (react-native-firebase)
- Android is receiving events from `onMessage` (react-native-firebase), but isn't receiving events from `onNotification` (react-native-push-notification)

**Comment by GioLogist** [comment]
That being said, the data is slightly different between platforms, even with the above implementation

Here's an example

```js
// admin SDK, triggering the push notification

const FCMToken = '' // edit
  admin.messaging()
  .sendToDevice(FCMToken, {
    notification: { 
      "title": "Hello",
      "body": "world", 
    },
    data: {
      "some": 'data'
    },
  })
  .then(function(response) {
    console.log('Successfully sent message:', response)
    process.exit()
  })
  .catch(function(error) {
    console.log('Error sending message:', error)
    process.exit()
  })
```


Log on iOS
```js
{
  "alert": { "body": "world", "title": "Hello" },
  "badge": undefined,
  "data": {
    "gcm.message_id": "1576292248612495",
    "google.c.a.e": "1",
    "notificationId": "E59A4076-FD49-4101-886C-EECAA4116339",
    "remote": true,
    "some": "data"
  },
  "finish": [Function finish],
  "foreground": true,
  "message": { "body": "world", "title": "Hello" },
  "sound": undefined,
  "userInteraction": false
}
```

Log on Android

```js
{
  "finish": [Function finish],
  "foreground": true,
  "id": "2040348668",
  "message": "world",
  "some": "data",
  "title": "Hello",
  "userInteraction": false
}
```

This makes it a bit more nuance to deal with. But at least somewhat functional. It seems as though w/Android, `notification.data` is merged into the top level. Even then, with some missing params. `remote` is omitted.

**Comment by GioLogist** [comment]
That being said, the data is slightly different between platforms, even with the above implementation

Here's an example

```js
// admin SDK, triggering the push notification

admin = require("firebase-admin")

const FCMToken = '' // edit
  admin.messaging()
  .sendToDevice(FCMToken, {
    notification: { 
      "title": "Hello",
      "body": "world", 
    },
    data: {
      "some": 'data'
    },
  })
  .then(function(response) {
    console.log('Successfully sent message:', response)
    process.exit()
  })
  .catch(function(error) {
    console.log('Error sending message:', error)
    process.exit()
  })
```

```js
// listener in react native

import PushNotification from 'react-native-push-notification'
import PushNotificationIOS from "@react-native-community/push-notification-ios";

export function initPushNotifications(){
  PushNotification.configure({
    // (optional) Called when Token is generated (iOS and Android)
    onRegister: function(token) {
      console.log("TOKEN:", token);
    },
  
  
    // (required) Called when a remote or local notification is opened or received
    onNotification: function(notification) {
      console.log("NOTIFICATION:", notification);

      // required on iOS only (see fetchCompletionHandler docs: https://github.com/react-native-community/react-native-push-notification-ios)
      notification.finish(PushNotificationIOS.FetchResult.NoData);
    },
    requestPermissions: false,
    senderID: '', 
  })
}
```


Log on iOS
```js
{
  "alert": { "body": "world", "title": "Hello" },
  "badge": undefined,
  "data": {
    "gcm.message_id": "---",
    "google.c.a.e": "1",
    "notificationId": "---",
    "remote": true,
    "some": "data"
  },
  "finish": [Function finish],
  "foreground": true,
  "message": { "body": "world", "title": "Hello" },
  "sound": undefined,
  "userInteraction": false
}
```

Log on Android

```js
{
  "finish": [Function finish],
  "foreground": true,
  "id": "---",
  "message": "world",
  "some": "data",
  "title": "Hello",
  "userInteraction": false
}
```

This makes it a bit more nuance to deal with. But at least somewhat functional. It seems as though w/Android, `notification.data` is merged into the top level. Even then, with some missing params. `remote` is omitted.

**Comment by GioLogist** [comment]
@alex-mironov I don't think this is yet stale?

**Comment by khlling** [comment]
I was looking into a similar issue. I was playing with both RNFB and RNPN and realised this through my own experimentation:

> In other words, I think you should stop trying to get RNPN to catch cloud messages. You just want to use it for it's local device notification API coverage. That's what RNFBv6 dropped - it doesn't handle local device notifications. But it has cloud messaging covered.

May be I was being blind but I couldn't see anything in the documentation that suggested this. Happy to update the documentation if we think that's useful?

I also found this too:

> Android is receiving events from onMessage (react-native-firebase)

But I found that RNPN was receiving events from onNotification leading to two notifications being triggered one from RNFB and one from RNPN. Interestingly I had to add these lines to make the RNFB notification look correct.

```
<meta-data android:name="com.google.firebase.messaging.default_notification_icon"
    android:resource="@mipmap/ic_notification" />
<meta-data android:name="com.google.firebase.messaging.default_notification_color"
    android:resource="@color/primary" />
```

I haven't tried this, but will give it a go: 

> iOS is receiving events from onNotification (react-native-push-notification) but not from onMessage (react-native-firebase)`

It seems that we have to write some platform specific code to deal with cross-platform FCM support?

**Comment by mikehardy** [comment]
I think this depends on what you send. If you mix notifications and data payloads in the cloud messages from firebase then things can get mixed up pretty quickly.

If you use data-only (silent push) from the cloud, handle it with onMessage in react-native-firebase and only use react-native-push-notifications to deal with the notifications display and click handling than it probably becomes a lot more clear.

Unfortunately there is one case that is not handled well by data-only firebase cloud messages and that is when an iOS app has been force-closed. Then the messages do not go through unless the app is restarted. The not-great workaround for this is to use the apns package from npm to directly send data-only messages to your iOS users which should wake the app so you can get them, then again use react-native-push-notifications to handle display etc

That is the theory anyway

**Comment by mstankov** [comment]
Hi @mikehardy 

First of all, thanks for all the support and hard work!

Regarding this:
> Unfortunately there is one case that is not handled well by data-only firebase cloud messages and that is when an iOS app has been force-closed. Then the messages do not go through unless the app is restarted. The not-great workaround for this is to use the apns package from npm to directly send data-only messages to your iOS users which should wake the app so you can get them, then again use react-native-push-notifications to handle display etc

I've been using RNFBv5 for data-only messages (silent notifications). As you mentioned - the major issue with it is that on iOS, once the app is killed or the user simply hasn't started it after a reboot, the messages don't wake up the app like in Android with a headlessjs task.
Initially I went for push notifications but my business case requires me to translate the body and title of the notifications in a language selected from the app. (stored in AsyncStorage). I failed to intercept the push notifications before they are displayed on iOS and Android in each case (foreground, background, killed) and after a lot of fruitless hours - decided to switch to data-only messages instead and capture them in the javascript side.

You mention an **apns package from npm** that provides the ability to directly send the messages. Would it be useful to me in my case or should I go back to push-notifications and try to intercept them again and translate the title/body on the native side? I'd really use a hint or two as to where to look for the messages in both Android and iOS before they are displayed in the status bar for each case!

In any case, thanks again and keep up the good work!

Regards

**Comment by mikehardy** [comment]
Hi @mstankov - if you send a payload that contains notification entries, the underlying firebase SDKs do all the work, you won't find the code and cannot intercept I don't think. So if you have to do I18N, and want to handle the iOS force-close case, I think you're on the right track with the missing piece being something I have not done myself but researched for a different query - maybe it's interesting? https://github.com/invertase/react-native-firebase/issues/3095#issuecomment-573805165

As mentioned there and I repeat here - I have not done this yet myself but I can't think of another way to handle all the cases everywhere including iOS app not getting messages after force-close - so I'm curious to hear anyone's experience and @7patricia might be as well

**Comment by maikmacs** [comment]
Hi @alex-mironov  did you find any solution? 

I have the same issue,  on IOS all it's Working,  On  Android Scheduled Notifications are working but Cloud Messaging only works in the test with the FCM , but when I send the Cloud Messaging only IOS are getting the notification

**Comment by JeffGuKang** [comment]
I set the `AndroidManifest` exactly same with @GioLogist's [thing](https://github.com/invertase/react-native-firebase/issues/2997#issuecomment-565673969) for android. And did not apply other RNPN's [manual](https://github.com/zo0r/react-native-push-notification#android-manual-installation). 

Before that, fcm token did not work as @alex-mironov. Only device token worked well in Android with RNPN. 

And reinstall application on android after deletion was better for notification test if I change `AndroidManifest.xml`. 

I hope this helps someone. [AndroidManifest](https://gist.github.com/JeffGuKang/d12796df10cfb22b2f00bc9cc98a4e02)

**Comment by bogrees** [comment]
Thank you @JeffGuKang and @alex-mironov!  I was able to **_finally_** get this working per your help.  I removed the following from the `AndroidManifest.xml`, and behold, cloud messaging push notifications are working again.

**REMOVED THIS:**
```
    <receiver android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationPublisher" />
    <receiver android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationBootEventReceiver">
        <intent-filter>
            <action android:name="android.intent.action.BOOT_COMPLETED" />
        </intent-filter>
    </receiver>
    <service android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationRegistrationService"/>
```

**KEPT THIS:**
```
    <service android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationListenerService"
      android:exported="false" >
      <intent-filter>
         <action android:name="com.google.firebase.MESSAGING_EVENT" />
      </intent-filter>
    </service>
```

**Comment by santinucera** [comment]
Hi, with the `AndroidManifest.xml` that @bogrees shows us, the Push Notifications from Firebase Cloud Messaging are working fine. But previously I was receiving Push Notifications from SNS and it stop  working. 
Is there any configuration that allows to have the Push Notification Services working together?

**Comment by bogrees** [comment]
@santinucera , more specifically, this seems to be the offending line (or more like a switch between the two):
```
<service android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationRegistrationService"/>
```
If that line is omitted, Push Notifications from Firebase Cloud Messaging work --- if it is included, then the latter notifications work.  Not sure what the fix for this is, but it does seem to be concentrated with that line specifically.

**Comment by santinucera** [comment]
@alex-mironov @mikehardy @GioLogist anyone has a solution to this problem?
I have been trying to find a solution for hours but I didn’t get any result. I’m not a specialist in Android’s configurations, so any help is welcome.

**Comment by harsh-syook** [comment]
i tried for hours to integrate the react native firebase v6 with react native push notifications to leverage the push notifications and finally i was able to get push notifications on my real android device. :)
now i am wondering where to handle onRefreshToken in react native push notifications library cos there is no method for that only onregister is there . but for now everything works

**Comment by alemkhodadadi** [comment]
I spent about 4 days and didn't find a complete solution for handling using both RNFB and RNPN. 

this is what i did:

1- I installed react-native-push-notification and get "senderId" from firebase console.
then I could receiving notifications. at this stage both onNotification and onRegister where working perfect. 
2- then i installed @react-native-firebase/app(or messeging or ...) for using analytics and crashlytics and the rediculous thing was that I couldn't receive notifications that I sent by using my packageName! I can now receive just notifications that I send by using token! 

I dont know what is happening .
here is my manifest:
`
<receiver android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationPublisher" />
  <receiver android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationBootEventReceiver">
      <intent-filter>
          <action android:name="android.intent.action.BOOT_COMPLETED" />
      </intent-filter>
  </receiver>
  <service android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationRegistrationService"/>
  <service
      android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationListenerService"
      android:exported="false" >
      <intent-filter>
          <action android:name="com.google.firebase.MESSAGING_EVENT" />
      </intent-filter>
  </service>
`

**Comment by JeffGuKang** [comment]
I updated android manifest for schedule notification. 

AndroidManifest.xml

```
    <!-- < Only if you're using GCM or localNotificationSchedule() > -->
    <uses-permission android:name="android.permission.WAKE_LOCK" />
    <permission
        android:name="${applicationId}.permission.C2D_MESSAGE"
        android:protectionLevel="signature" />
    <uses-permission android:name="${applicationId}.permission.C2D_MESSAGE" />
    <!-- < Only if you're using GCM or localNotificationSchedule() > -->
```

In `<application> </application>`
```
<!-- < Only if you're using GCM or localNotificationSchedule() > -->
        <receiver
            android:name="com.google.android.gms.gcm.GcmReceiver"
            android:exported="true"
            android:permission="com.google.android.c2dm.permission.SEND" >
            <intent-filter>
                <action android:name="com.google.android.c2dm.intent.RECEIVE" />
                <category android:name="${applicationId}" />
            </intent-filter>
        </receiver>
        <!-- < Only if you're using GCM or localNotificationSchedule() > -->
        
        <receiver android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationPublisher" />
        <receiver android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationBootEventReceiver">
            <intent-filter>
                <action android:name="android.intent.action.BOOT_COMPLETED" />
            </intent-filter>
        </receiver>
        <service android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationRegistrationService"/>
        
      <!-- < Only if you're using GCM or localNotificationSchedule() > -->
        <service
            android:name="com.dieam.reactnativepushnotification.modules.RNPushNotificationListenerServiceGcm"
            android:exported="false" >
            <intent-filter>
                <action android:name="com.google.android.c2dm.intent.RECEIVE" />
            </intent-filter>
        </service>
```

**Comment by Ehesp** [comment]
I've just scanned over this issue and I'm still a bit confused as to what the actual issue is :D I'm working through Messaging at the moment (https://github.com/invertase/react-native-firebase/pull/3309) to give it some love, however I'm not 100% what the actual problem is here?

**Comment by Salakar** [comment]
#3339 is now merged and the latest RC, 6.4.0-rc4 is up on NPM - so I'm going to go ahead and close this issue as that PR should solve all outstanding issues relating to not receiving messages, see the PR description for fixes.

**Comment by fspider** [comment]
Hello. I am having problem with ios.
1) Firebase Cloud Messaging -> Android (React-native-push-notification) [WORKING WELL]
2) Firebase Cloud Messaging -> APN -> IOS (React-native-push-notification) [Not WORKING]
3) Knuff opensource test app(MAC) -> APN -> IOS (React-native-push-notification) [WORKING]

Problem is in case of 2)
Here, i can get registration token like "6f6fd4c98cf3e4de8f427a48ae196859b70e2a3b6680715709f52fd44a737672"
But can not receive firebase cloud message.

I have implemented all for "React-native-push-notification", "@react-native-community/push-notification-ios"
But not sure for firebase-messaging.


What do i have to do more to integrate with firebase cloud messaging?

**Comment by mikehardy** [comment]
Troubleshooting messaging without complete explicit control of the payload is a real timekiller.
Use a script to post directly so you can control the JSON, then test things and if you problems post the exact JSON you are using here. For example:

```bash

(base) mike@kunashir:~/work/Kullki/ksocialscore/packages (rebranding *) % cat public-app/publish/test-silent-push.sh 
#!/bin/bash

PROD_SERVER_KEY="MY BIG SUPER SECRET PROD API KEY HERE"

SERVER_KEY=$PROD_SERVER_KEY
#SERVER_KEY=$DEV_SERVER_KEY

curl -X POST \
--header "Authorization: key=$SERVER_KEY"  \
--Header "Content-Type: application/json" https://fcm.googleapis.com/fcm/send -d "{\"to\":\"/topics/AppInfo\",\"collapse_key\": \"AppInfo\",\"content_available\": true, \"priority\": \"high\", \"data\":{\"IconCheck\":\"Testing Icon\"}}"


```

**Comment by fspider** [comment]
I am using this code in PHP server to send push notification through firebase.
Problem is that IOS can't receive PN. 
While PN is working on Android.

```
		$device_token = json_decode(json_encode($device_token));
		$notification = array('title' => $title, 'body' => $body, 'sound' => 'default', 'badge' => '1',  'content-available' => '1');
		$fields = array('registration_ids' => $device_token, 'notification' => $notification, 'priority' => 'high', 'data' => $data,  'content-available' => '1');
		$args = array(
		  'timeout'   => 45,
		  'redirection' => 5,
		  'httpversion' => '1.1',
		  'method'    => 'POST',
		  'body'      => json_encode($fields),
		  'sslverify'     => false,
		  'headers'     => array(
			'Content-Type' => 'application/json',
			'Authorization' => 'key=' . $FIREBASE_API_KEY,
		  ),
		  'cookies'     => array()
		);
		$response = wp_remote_post('https://fcm.googleapis.com/fcm/send', $args);
```


Here is one more question.
I can get two defferent types of token on ios:

This is when registering with react-native-push-notification 
"6f6fd4c98cf3e4de8f427a48ae196859b70e2a3b6680715709f52fd44a737672"

Another one is when with firebase
dxTqEUQyV0yKuacjOzyCEB:APA91bF1kmK0jGtxR6bssgI8ANOoHfd5dqoXQhAiMmtHtUNGJTTRhERXqbTqEnyi7z1OadDyDTtqkFTXfZwoCC_BEBoHjNEC9xreDvO

Which one is correct token for sending firebase cloud message?

**Comment by fspider** [comment]
This is exact log history for token from FIrebaseMessaging and RNPN

> 2020-04-12 13:31:03.868830+0800 RPGAssassin[12015:1630030] -[RNFirebaseMessaging messaging:didReceiveRegistrationToken:] [Line 86] Received new FCM token: cc5WKhOZckvNlMvgqzevF3:APA91bFodWWjQ3aViz5uCOulOizhGbr_wbOgn1rrIpkp9rHpU5J6-TvXyLY9Nlv1V_d0DS7IO3arBZwK0aLtJtZin_FTWGV4I-BgRyRekiQOO_kiY69b14eumhO6iMgBH-2d7AU9BWnN
> 2020-04-12 13:31:17.565 [info][tid:com.facebook.react.JavaScript] 'Registere TOKEN:', '6f6fd4c98cf3e4de8f427a48ae196859b70e2a3b6680715709f52fd44a737672'

**Comment by fspider** [comment]
Is it impossible to handle firebase cloud message with react-native-push-notification?

**Comment by leemcmullen** [comment]
@fspider Hey, to answer your question (https://github.com/invertase/react-native-firebase/issues/2997#issuecomment-612567047) the correct token is the longer one you mentioned i.e. dxTqEUQyV0yKuacjOzyCEB:APA91bF1kmK0jGtxR6bssgI8ANOoHfd5dqoXQhAiMmtHtUNGJTTRhERXqbTqEnyi7z1OadDyDTtqkFTXfZwoCC_BEBoHjNEC9xreDvO

I know you're having issues when sending from PHP, have you tried sending directly from the Firebase console testing UI? Just thinkiing that would at least remove a potential issue.

Also do you have a multi scheme/configuration environment for your iOS app? The reason I ask is because I had the situation where I wasn't getting iOS PNs coming through for my UAT instance of the app but I was for my DEV instance. It turns out the reason was that I'd simply forgotten to add the APNS auth key into the Firebase console for the UAT app 🤦

**Comment by leemcmullen** [comment]
Hey @mikehardy, thanks for your input on this thread, very helpful. One question is whether it's possible to get react-native-push-notification (to **only** show local notifications) working alongside RNFB (which is handling remote/cloud notifications)?

For my specific case, I have RNFB handling all remote notifications from FCM and that's working well for both iOS and Android. All I want to be able to do is occasionally show a local notification (which I understand RNFB dropped with v6). 

You mention this case in your comment here (https://github.com/invertase/react-native-firebase/issues/2997#issuecomment-565086660). So I just wanted to double double check that the scenario you outlined is in fact still possible before I end up down some 3 day rabbit hole only to find that it's not actually possible! Thanks 😄

**Comment by fspider** [comment]
Thanks for your answer!

Have you tried sending directly from the Firebase console testing UI? Yes,Using Firebase could messaging test UI
Also do you have a multi scheme/configuration environment for your iOS app? Sorry, I don't understand. Was testing with debug/Release mode in original scheme.
And already added APNS auth key on FIrebase configuration.


By the way, I solve this problem like this.
For IOS, I sent PN directly to APN server.
It works.
PHP server -> FIrebase ->Android
PHP Server -> APN -> IOS

**Comment by mikehardy** [comment]
Yeah, people have reported success with all of react-native-notifications, react-native-push-notifications, and notifee
The setup for any of these things is VERY DIFFICULT. It is just a difficult area full of irreducible complexity. Be prepared to integrate very very slowly testing each app state and feature. But they all have reported success.

**Comment by rishiankush** [comment]
Is there any proper example for integrating the latest firebase version with react native 0.62 or above 0.60. I found nothing useful and getting error after following all the instruction in the latest documentation provided here:https://rnfirebase.io/
The error that I am getting is:
<img width="1238" alt="Screenshot 2020-04-25 at 12 22 05 PM" src="https://user-images.githubusercontent.com/25289937/80273411-6b140000-86ef-11ea-8743-6f51305c5a94.png">

**Comment by vinayjags** [comment]
@alex-mironov Hello, I was able to successfully integrate with Android, but for iOS i am still struggling. Can you please provide me some pointers.
